src/libsystemd/sd-path/sd-path.c

```diff
@@ -98,10 +99,17 @@ static int from_user_dir(const char *field, char **buffer, const char **ret) {
         assert(buffer);
         assert(ret);
 
-        r = from_home_dir(NULL, ".config/user-dirs.dirs", &b, &fn);
+        r = from_home_dir("XDG_CONFIG_HOME", ".config", &b, &c);
         if (r < 0)
                 return r;
 
+        if (endswith(c, "/"))
+                fn = strappend(c, "user-dirs.dirs");
+        else
+                fn = strjoin(c, "/", "user-dirs.dirs", NULL);
+        if (!fn)
```

Comment:
You don't actually need to check for the slash here. We only do that when we use something for output, and `fn` is not printed anywhere. So just append "/user-dirs.dirs" always.

Second problem is that `fn` is leaked. It's probably best to replace `strappend` with `strappenda` which allocates on the stack and freeing is automatic.